library/OrderCloud.Catalyst/DataMovement/ListAllAsync/ListAllHelper.cs

```diff
@@ -0,0 +1,139 @@
+		{
+			var pageTasks = new List<Task<ListPage<T>>>();
+			var totalPages = page1.Meta.TotalPages;
+			var page = 2;
+			while (page <= totalPages)
+			{
+				pageTasks.Add(listFunc(page++));
+				var running = pageTasks.Where(t => !t.IsCompleted && !t.IsFaulted).ToList();
+				if (!running.Any()) { continue; }
+				if (totalPages == 0 || running.Count >= MAX_NUM_PARALLEL_REQUESTS) // throttle parallel tasks
+					totalPages = (await await Task.WhenAny(running)).Meta.TotalPages;  //Set total number of pages based on returned Meta.
+			};
+			var data = (
+				from finalResult in await Task.WhenAll(pageTasks) //When all pageTasks are complete, save items in data variable.
+				from item in finalResult.Items
+				select item).ToList();
+			data.AddRange(page1.Items);
+			return data;
+		}
+
+		private static async Task<List<T>> ListAllWithFacetsByFilterAsync<T>(ListPageWithFacets<T> page1, PropertyInfo id, Func<KeyValuePair<string, object>, Task<ListPageWithFacets<T>>> listFunc)
+		{
+			var data = page1.Items as List<T>;
+			var filter = IGNORED_FILTER;
+			var page = 2;
+			while (page <= page1.Meta.TotalPages)
+			{
+				var response = await listFunc(filter);
+				data.AddRange(response.Items);
+				var lastID = id.GetValue(response.Items.Last()) as string;
+				filter = new KeyValuePair<string, object>("ID", $">{lastID}");
+				page++;
+			}
+			return data;
+		}
+
+		private static async Task<List<T>> ListAllWithFacetsByPage<T>(ListPageWithFacets<T> page1, Func<int, Task<ListPageWithFacets<T>>> listFunc)
+		{
+			var pageTasks = new List<Task<ListPageWithFacets<T>>>();
+			var totalPages = page1.Meta.TotalPages;
+			var page = 2;
+			while (page <= totalPages)
+			{
+				pageTasks.Add(listFunc(page++));
+				var running = pageTasks.Where(t => !t.IsCompleted && !t.IsFaulted).ToList();
+				if (!running.Any()) { continue; }
+				if (totalPages == 0 || running.Count >= MAX_NUM_PARALLEL_REQUESTS) // throttle parallel tasks
+					totalPages = (await await Task.WhenAny(running)).Meta.TotalPages;  //Set total number of pages based on returned Meta.
+			};
+			var data = (
+				from finalResult in await Task.WhenAll(pageTasks) //When all pageTasks are complete, save items in data variable.
+				from item in finalResult.Items
+				select item).ToList();
+			data.AddRange(page1.Items);
+			return data;
+		}
+
+
+		// See https://github.com/tmenier/Flurl/blob/ce480aa1aa8ce1f2ff4ebce9f1d6eaf30b7d6d8c/src/Flurl.Http/Configuration/DefaultUrlEncodedSerializer.cs
+		public static string AndFilter(this object filters, KeyValuePair<string, object> filter)
+		{
+			var filterList = filters?.ToKeyValuePairs().ToList() ?? new List<KeyValuePair<string, object>>();
+			filterList.Add(filter);
+			var qp = new QueryParamCollection();
+			foreach (var kv in filterList)
+			{
+				if (kv.Value != null)
+				{
+					qp.Add(kv.Key, kv.Value);
+				}
+			}
+			return qp.ToString(true);
+		}
+	}
+}
```

Comment:
Can you add summary comments for each method? That way if somebody is just using IntelliSense they will still get some documentation. Especially in this helper class since there are so many different ways to list